blog: How the controller-runtime Cache Actually Works (#55914)
kubernetes/website: content/en/blog/_posts/2026/controller-runtime-cache-explained/index.md

not risk missing a real event.

There is also `ctrl.Result{Requeue: true}` — enqueue immediately, subject to the rate
limiter.

## cache + index = almost SQL

Now you get to what is, arguably, the most useful capability of the cache — and the one most
operators leave unused.

By default, a `List` from the cache looks like this:

```go
var pods corev1.PodList
_ = r.List(ctx, &pods)
for _, p := range pods.Items {
    if p.Spec.NodeName == "node-1" {
        // do something
    }
}
```

It works — until the cluster has 50,000 Pods and reconciles run hundreds of times per
second, at which point the controller is shuffling the same half-gigabyte of pointers back
and forth on every trigger. `O(n)` per reconcile.

The Indexer in `client-go` can do much better. You declare up front which field you want to
index on:

```go
// Index by spec.nodeName for Pods
if err := mgr.GetFieldIndexer().IndexField(
    ctx,
    &corev1.Pod{},
    "spec.nodeName",
    func(obj client.Object) []string {
        pod := obj.(*corev1.Pod)
        if pod.Spec.NodeName == "" {
            return nil
        }
        return []string{pod.Spec.NodeName}
    },
); err != nil {
    return err
}
```

> 

Reviewer: The passage might make it sound like the field to be indexed is an actual JSONPath to a field in the object and its value must be the value of that field. That is not the case. I suggest pointing this out explicitly.

```suggestion

Two things about that call are worth making explicit, because the tidy example hides
them behind a convention.

**The index name is arbitrary.** That second argument, `"spec.nodeName"`, is just a
string key the index is registered under. `controller-runtime` does not parse it as
JSONPath and does not check it against the object's schema — you could write `"by-node"`
or `"xyzzy"` and it would behave identically. The only rule is that the *exact same
string* comes back in `MatchingFields` at query time. Naming the index after the field it
happens to read is a readability convention, nothing more.

**The indexed value is computed, not read.** The function returns whatever strings you
build; they need not be the verbatim contents of any single field. You can lowercase a
value, join several fields into one composite key, bucket a timestamp (the **time-bucket**
trick below does exactly this), or emit a string that appears nowhere in the object
literally. Whatever the function returns becomes a key in the inverted dictionary, and a
`MatchingFields` lookup for that exact key is what finds the objects again. The only constraint is that the value has to be *derivable from the object you are indexing*.

```

Author: Good catch, thanks. Applied as suggested, just re-wrapped a couple of lines to match the file.

---
layout: blog

> title: "How the controller-runtime cache actually works, and why your operator does not crush the API server"

Reviewer 2: the Kubernetes style guide recommends title case for the page title in front matter:
https://kubernetes.io/docs/contribute/style/style-guide/#headings

```suggestion
title: "`How the controller-runtime Cache Works, and Why Your Operator Does Not Crush the API Server"
```

Author: Thanks! Looking at the same /#headings section, the guidance is actually sentence case ("Use sentence-case capitalization for headings"), so I'd prefer to keep the current title. Happy to change it if there's a more specific blog-titles convention I missed.

Reviewer 2: > the guidance is actually sentence case ("Use sentence-case capitalization for headings")

It recommends to use sentence-case for headings in the page body, but for the title in the front matter, it recommends title-case.

<img width="635" height="221" alt="image" src="https://github.com/user-attachments/assets/6672ad3d-09b6-483e-880d-0825aff04a0f" />

Author: You're right, I missed that distinction. Updated, thanks!

> title: "How the controller-runtime Cache Actually Works, and Why Your Operator Does Not Crush the API Server"

Reviewer 3: Did you mean "crash" instead of "crush"?

Reviewer: Good catch, in this case either word would be appropriate :grin:

I'm not able to edit @kvaps' branch myself, but if you feel it'd be clearer, you're welcome to change `s/Crush/Crash/`.

Author: Good catch, thanks. Updated.

> title: "How the controller-runtime Cache Actually Works, and Why Your Operator Does Not Crash the API Server"

Reviewer 4: Would you folks be willing to change "Operator" to "Controller"?

A controller that doesn't crash the API server relies on the same qualities to achieve this, whether or not it is also an operator.

> date: 2026-06-01

Reviewer 2: The Kubernetes blog submission guidance says initial blog PRs should be merged as drafts and should not specify a publication date until scheduling. 
https://kubernetes.io/docs/contribute/blog/article-submission/#front-matter

```suggestion
draft:true
```

Author: Applied — replaced the date with `draft: true`; happy to set the publication date at scheduling time per the submission guidance.

slug: controller-runtime-cache-explained
author: >
  Andrei Kvapil (Ænix),
  Timofei Larkin (Ænix)
---

Kubernetes has long been the default platform for distributed workloads, and writing your own
operator for it is now a matter of a few hours. The standard path — `kubebuilder` on top of
`controller-runtime` — gives you a project scaffold, types, and a reconciler. For typical
scenarios that is more than enough. But as soon as load grows or the operator starts behaving
in ways you did not expect, a whole class of edge cases shows up. Most of them trace back to
the same root cause: a fuzzy mental model of how `controller-runtime` works inside. If you
write Kubernetes controllers, this article should help you build a coherent picture and avoid
expensive surprises in production.

This article walks through the internals of `controller-runtime` and, along the way, shows which
architectural decisions are baked into Kubernetes itself. The starting point is how
controllers actually read objects from the Kubernetes API.

> 

Reviewer 4: By this point, the article should either have mentioned Go / Golang, or be just about to?

Why? Well, you don't know that the reader is thinking in Golang terms. You can write a controller in any programming language and it is still a controller, but this library and its specific design is relevant mostly to people writing a controller in Go.

A common misconception goes like this: `r.Get()` inside `Reconcile` queries `kube-apiserver`
directly; `r.List()` returns a fresh, live view of the world; and after `r.Update()` you can
re-read the object and immediately see the new state. In practice the model is the opposite:
`controller-runtime` operates against a local copy of the data populated through `LIST+WATCH`.
Reads inside a reconciler cost almost nothing and do not load the control plane even at
hundreds of calls per second — but the price of this design is that an operator can quietly
consume gigabytes of memory, perform hidden `O(n)` scans, and regularly trip over stale reads.

This post is aimed at engineers who already write operators in Go with `controller-runtime`
but want to consolidate the pieces into a single mental model rather than carry around a bag
of isolated observations. The focus is the practical impact on production clusters: memory,
network traffic, read consistency, and reconciler behavior.

## TL;DR

If you take only one idea from this article, take this:

`r.Get()` and `r.List()` inside a reconciler typically do not read from the API server. They

> read from a local in-memory cache, which the manager warms up with `LIST` and then keeps

Reviewer 4: nit: logical verbs should be formatted in lowercase bold; **list** not `LIST`.

  of objects.
- `APIReader` is rarely needed — but in some places you really cannot do without it.

The rest of the article unpacks why this is so and how the model is wired underneath.

## A bit of context: what a reconciliation loop is

To avoid arguments about terminology, start with the basic model.

A controller in Kubernetes lives inside a reconciliation loop: it continuously compares the
desired state of an object with the actual state and tries to bring one in line with the
other. The idea is described in the original
[architectural notes](https://github.com/kubernetes/design-proposals-archive/blob/main/architecture/principles.md)
on Kubernetes. In practice it looks like this:

- A user or another controller mutates an object.
- An event lands in a queue.
- `Reconcile` reads the current state.
- The controller decides what to create, update, or delete.
- The system produces a new event and the loop repeats.

What matters here is not that the controller "does something" — it is **where it learns about
changes from** and **where it reads state from**. That is exactly where the cache comes in.

On a live cluster, the easiest way to see this in action is:

```bash

> kubectl get pod -w

Reviewer 4: Better:
```suggestion
kubectl get pods --watch
```

More to type, but easier to understand.

```

In `-w` (watch) mode `kubectl` subscribes to the same event stream that controllers consume.
You create or delete a Pod and you see not a single "final" object but a chain of states: the
scheduler assigns a node, the kubelet updates status, other controllers contribute their
changes. Kubernetes controllers do not poll continuously — they consume an event stream and
maintain a local state that is kept current.

For a visual walkthrough, see [Reconciliation loop pattern in visual representation](https://www.youtube.com/watch?v=P50otWVh7w4),
a talk that shows how the reconciliation loop plays out on a real Pod and the states it
passes through.

## Why the cache exists in `controller-runtime` at all

Imagine the simplest possible controller:

```go
func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
    var pod corev1.Pod
    if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
        return ctrl.Result{}, err
    }
    // ... meaningful logic ...
}
```

Looks straightforward. But what happens when you call `r.Get`? Does it fire an HTTP request at
the API server? If it did, picture the scene: a dozen operators, each running a few

> controllers, each issuing a `Get` and a `List` per reconcile, with hundreds of reconciles per

Reviewer 4: Maybe you meant the API verbs here? Not sure, but if you meant the API verbs, try lowercase bold.

second. The API server and `etcd` would be writing each other farewell letters within minutes.

> To prevent that, Kubernetes was built around a **watch model** rather than polling from the

Reviewer 4: ```suggestion
To prevent that, Kubernetes was built around a _watch model_ rather than polling from the
```

> very beginning. The standard mechanism works like this: a client issues `LIST` once, gets a

Reviewer 4: nit: **list**

snapshot of the slice of the world it cares about, then subscribes to a stream of changes via

> `WATCH` and keeps a local copy current. Everything happens over a single long-lived HTTP

Reviewer 4: nit: **watch**

connection, with no "what is in the world right now?" loop.

This idea has lived in `client-go` since the very first controllers in
`kube-controller-manager`. `controller-runtime` wraps it in a friendly framework so
that you do not have to glue `Reflector`, `DeltaFIFO`, and `Indexer` together yourself (more
on those below).

So when people talk about "the controller-runtime cache", they are not talking about a clever
optimization. They are describing the foundation of the entire model: you read from memory,
you write to the API server, and you receive feedback through a watch.

The rest of this article walks through how each piece is wired up.

## Glossary

A few terms collected up front, so you do not have to jump back and forth later. Skim or skip
if any of them are already familiar.

- **GVK (GroupVersionKind)** — the triple that uniquely identifies a resource type in
  Kubernetes: group, version, and kind, for example `apps/v1/Deployment`. Almost every API in

>   `controller-runtime` works in terms of GVK rather than "the resource name as it appears in

Reviewer 4: I might say API name or API type. If you have a Deployment named myapp, myapp is a resource. It's true that `deployments` is also a resource (it's the collection), but I think this detail confuses people more than it helps.

  kubectl".

> - **resourceVersion** — an opaque string the API server attaches to every object (under the

Reviewer 4: Not opaque (since v1.35); check the current docs, and maybe hyperlink there because the nuance isn't easy to gloss.

Author: Good catch, you're right. Updated the glossary — dropped the flat "opaque" wording and noted that since 1.35 resource versions are orderable within a type (it's a conformance requirement now), with a link to api-concepts.

Also, while digging I noticed the `ObjectMeta.ResourceVersion` godoc still has a leftover broken line ("…opaque by clients and ."), so I sent a small fix upstream: kubernetes/kubernetes#139969

  Each informer in `controller-runtime` has its own store.

- **ResourceEventHandler** — an interface with three methods: `OnAdd`, `OnUpdate`, `OnDelete`.
  The informer calls them for every event delivered through DeltaFIFO. The store is updated
  in lockstep with the handler invocation, so a handler already sees the latest version of
  the object in the indexer. Subscribers (your controllers) register handlers like this and
  learn about changes through them.

- **workqueue** — a queue of object keys (`namespace/name`) with deduplication and rate
  limiting. On every event the controller enqueues a key; workers pop keys one at a time and
  hand them to `Reconcile` as a `ctrl.Request`.

- **Predicate** — a controller-side filter. A predicate decides whether an event should be
  enqueued at all (for example, "react only to changes in `spec`, ignore `status`").

With those in hand, you can dive in.

## Anatomy: what lives under the cache package

If you peek into `sigs.k8s.io/controller-runtime/pkg/cache`, you will see that it is a thin
wrapper over `k8s.io/client-go/tools/cache`. The same primitives that power the rest of
Kubernetes live underneath:

> - **Reflector** — keeps a `WATCH` open against the API server and writes incoming changes

Reviewer 4: nit: **watch**

>   into a queue as **deltas**. A delta is a record of the form "object X received an `Added` /

Reviewer 4: nit: _deltas_ (italic)

  `Updated` / `Deleted` event, and here is its new version". Effectively a single line in a
  change log.

- **DeltaFIFO** — the queue that holds those deltas. Per `namespace/name` key it accumulates
  the list of things that happened to that object, in order.

- **Indexer (Store)** — the in-memory object store, plus the indexes built over it.

- **SharedIndexInformer** — the conductor that ties everything together and dispatches events
  to subscribers — your controllers and any other observers.

At a glance the pipeline looks like this:

> ![Pipeline diagram: API server → Reflector → DeltaFIFO → Indexer → Event handlers](pipeline.svg)

Reviewer 2: The alt text is helpful, and SVG is a good fit here. One small blog-guidelines suggestion: for diagrams, the Kubernetes blog guidance recommends using the figure shortcode where feasible.
https://kubernetes.io/docs/contribute/blog/guidelines/#illustrations

```suggestion
{{< figure src="pipeline.svg" alt="Pipeline diagram: API server to Reflector to DeltaFIFO to Indexer to Event handlers" >}}
```

Author: Applied, thanks — switched all three diagrams to the `figure` shortcode.

Now walk through each link.

### Reflector and resourceVersion

The Reflector is the only component that talks to the API server directly. It has exactly two

> jobs: do a single `LIST` at startup, then keep a `WATCH` open from there on.

Reviewer 4: nit: **list** and **watch**

This is where the `resourceVersion` earns its keep. Along with the list of objects, the API
server returns the version at which the snapshot was produced. The Reflector then says to the

> API server, "open a `WATCH` from version X", and receives a stream of events for everything

Reviewer 4: nit: **watch**

that happened after that version. That is the basis of consistency: there is no risk of
missing an event between `LIST` and `WATCH`, because `WATCH` resumes exactly at the point
where `LIST` ended.

If the connection drops, the Reflector reconnects with the last known `resourceVersion`. If
the API server replies with `410 Gone` ("that version is no longer in the history, you are

> too far behind"), the Reflector performs a fresh `LIST` and starts over. This is called a

Reviewer 4: nit: **list**

> **relist**, and it does not happen on a schedule — only in those failure scenarios.

Reviewer 4: nit: _relist_ (italics)

### DeltaFIFO: a queue of deltas

> This piece is worth pausing on. DeltaFIFO is the buffer between the Reflector and the rest of

Reviewer 4: I would put DeltaFIFO inside backticks.

the informer. Its input is a stream of events from the API server; its output is the same

> events, but **grouped by key** and in strict order.

Reviewer 4: not: use italics not bold

More precisely, DeltaFIFO solves three problems:

1. **It preserves order.** Whatever stream of changes flows in for `default/my-deploy`, the
   consumer sees the same ordering the API server delivered.
2. **It groups by key.** All deltas for a single `namespace/name` accumulate in one slot.
   `Pop()` returns not a single delta but a **slice** of every delta accumulated under that
   key — the consumer sees, in one shot, everything that has happened to the object since the
   last call.
3. **It deduplicates selectively.** The built-in `dedupDeltas` function collapses
   **consecutive `Deleted` deltas** for the same key, so two delete events do not turn into
   two separate processing rounds.

> An important caveat: **DeltaFIFO does not merge consecutive `Added` or consecutive `Updated`

Reviewer 4: (bold here is fine; no need to change it)

deltas.** Collapsing every intermediate state into a single final one is, in general, not its
job.

A worked example. Suppose three events for object `default/my-deploy` arrive in quick
succession:

1. `Added` — the Deployment is created (say, with `spec.replicas=1`).
2. `Updated` — somebody bumps `spec.replicas` to `2`.
3. `Updated` — and immediately to `3`.

DeltaFIFO places all three deltas into the slot keyed by `default/my-deploy`. `Pop()` returns
them as a single slice, and `sharedIndexInformer.HandleDeltas` walks through them in order:
first `OnAdd`, then two `OnUpdate` calls (one for the intermediate `1→2` transition and one
for the final `2→3`). The event handler runs three times, no shortcuts.

There **is** per-object deduplication, but not in DeltaFIFO — it lives one layer up, in the
controller's workqueue. The mechanic is straightforward: for each delta from DeltaFIFO, the

> controller's event handler extracts the `namespace/name` **key** from the object and enqueues

Reviewer 4: nit: _key_ (in italics)

Yes — at heart it is a map in memory. No B-trees, no LSMs. That is precisely why a cache-hit
`r.Get` costs microseconds: it is a map lookup followed by a copy of a Go struct.

### SharedIndexInformer and subscriptions

A SharedIndexInformer fuses Reflector, DeltaFIFO, and Indexer together and exposes two
interfaces to the rest of the world:

- Read objects directly from the indexer.
- Register a `ResourceEventHandler` and receive notifications for every event coming out of
  DeltaFIFO — `OnAdd`, `OnUpdate`, `OnDelete`. The store is updated in lockstep with the
  handler call, so by the time your handler runs, the indexer already reflects the new state.

"Outside" here means your controllers. When a controller registers `Watches(...)`, under the
hood it asks the informer: "add a handler that, on every change, enqueues the key into my
workqueue". The controller's workers then pop keys one at a time and call your
`Reconcile(ctx, ctrl.Request{NamespacedName: ...})`.

The keyword in the name is **Shared**. The manager creates **one** informer per GVK, and
every controller, webhook, and event source within that manager subscribes to it:

> ![Shared informer diagram: a single LIST/WATCH per GVK feeding multiple subscribers](shared-informer.svg)

Reviewer 2: Same here.
```suggestion
{{< figure src="shared-informer.svg" alt="Shared informer diagram: a single LIST/WATCH per GVK feeding multiple subscribers" >}}
```

Author: Applied, thanks.

In other words: an informer is the thing that subscribed to Pods once, holds them locally,
and serves every interested party in the process. From the API server's perspective, that is

> one `LIST` and one `WATCH` per GVK, regardless of how many reconcilers live inside your

Reviewer 4: nit: use **list** and **watch**

4. After that, a `WATCH` is opened starting from the `resourceVersion` returned by `LIST`.
5. **Only then** does the controller start invoking `Reconcile` — specifically, once
   `cache.WaitForCacheSync` has returned `true` for every source it owns. Until that point,
   workers do not drain the workqueue, even if events have already started piling up.

So in `controller-runtime`, "the reconciler is running but the cache is still empty" is
**not a state you can ever observe** by construction. The warm-up always happens up front,
never lazily.

What happens during the first `r.Get`? Suppose your reconciler contains:

```go
var obj appsv1.Deployment
err := r.Get(ctx, req.NamespacedName, &obj)
```

Under the hood it boils down to roughly this:

```go
item, exists, err := indexer.GetByKey("default/my-deploy")
if !exists {
    return apierrors.NewNotFound(...)
}
// DeepCopy into obj
```

No HTTP, no TLS, no protobuf serialization, no `etcd`. A map lookup, a struct copy, return.
Microseconds.

To repeat, because it matters: even the very first `Get` in the controller's lifetime reads
from a fully warmed-up, fully indexed snapshot. There is no "first time slow, then fast".

> > **Note:** This applies specifically to `mgr.GetClient()`. If for some reason you need to

Reviewer 4: Don't use a block quote here. Some Markdown dialects use special block quotes for callouts; we don't.

## Client ≠ Cache: read from memory, write to the API server

Another point that often gets lost. `client.Client` in `controller-runtime` is a composite
object:

- **Reads** (`Get`, `List`) go through the cache.
- **Writes** (`Create`, `Update`, `Patch`, `Delete`, `DeleteAllOf`) go straight to the API
  server.

This is not a hack — it is a deliberate design choice:

- Reads are frequent; they should be cheap.
- Writes are rare; they should be exact.
- Writing through the cache would invite split-brain — the local copy thinks the change went
  through, while the API server has already rejected it.

It is worth dwelling on "should be exact". This is where `resourceVersion` shows up again.

When you read an object from the cache, you do not get its current state in `etcd` — you get
the state as the Reflector last observed it. That state carries a `resourceVersion`. You then
mutate the object and call `r.Update(ctx, &obj)`. The request goes to the API server right
now, and the API server checks:

- Does the `resourceVersion` in your PUT match the `resourceVersion` in `etcd`? Yes — write
  it.
- No, `etcd` already has a newer one? Reply with `409 Conflict` — somebody beat you to it.

> This is **optimistic concurrency control**. No real locks are taken; everybody writes in

Reviewer 4: ```suggestion
This is _optimistic concurrency control_. No real locks are taken; everybody writes in
```

parallel; but only one of the racing `Update` calls wins — the one that arrives with the
current version. Everyone else gets a `409` and is expected to re-read and try again.

Why does this matter for the cache? If you naively send a PUT with "your" `resourceVersion`
from the cache and somebody has updated the object since you read it, you will get `409`.
That is **not a bug**. It is exactly the protection the system is supposed to give you.

> Writing without the `resourceVersion` check (via `Patch` without an optimistic lock, or via

Reviewer 4: nit: at this point, you could link to the docs on doing this

Server-Side Apply) is also possible, but that is a separate conversation.

The "write → visibility" cycle now looks like this:

> ![Write visibility diagram: client.Update → API server → watch event → cache](update-visibility.svg)

Reviewer 2: Same here.

```suggestion
{{< figure src="update-visibility.svg" alt="Write visibility diagram: client.Update to API server to watch event to cache" >}}
```

Author: Applied, thanks.

`pod.Labels["foo"] = "bar"` directly inside a handler. Historically, `Get` and `List`
returned a pointer into the store as well, with predictable consequences: somebody patched a
status "for convenience" in one controller and broke the world view of an unrelated
controller next door.

Today, `controller-runtime` performs a `DeepCopy` on `Get` and `List` by default. The simple
rule:

- Anything you receive from `r.Get` / `r.List` is yours; mutate freely.
- Anything you receive in a `Predicate` or an `EventHandler` is shared, not yours. If you
  must mutate it, call `obj.DeepCopy()` first; otherwise you are silently corrupting the
  cache for every other controller subscribed to the same type.

A concrete review heuristic: if `predicate.Funcs{UpdateFunc: ...}` or
`handler.EnqueueRequestsFromMapFunc(...)` contains expressions like
`e.ObjectNew.SetLabels(...)` or `obj.Status.X = Y`, stop and ask whether a `DeepCopy` is
missing before that mutation.

### Mistake 3: resync is not relist

An informer has a `resyncPeriod` parameter (10 hours by default in `controller-runtime`), and
many people read it as "rebuild the cache from the API server every N hours".

> It does not. **Resync does not perform a `LIST`.** It re-emits everything currently in the

Reviewer 4: ```suggestion
It does not. A resync does **not** perform a **list**. It _re-emits_ everything currently in the
```

indexer back through DeltaFIFO as `Sync` deltas, and the informer processes them as usual,
calling `OnUpdate(old, old)` for each object. This gives a controller that has somehow
missed its reconcile window (a stuck worker, a dropped handler) a chance to see the world
again. It generates no traffic to the API server.

> A real `relist` happens only in two cases: when the `WATCH` died with `410 Gone`, and when

Reviewer 4: nit: watch should be lowercase bold

index on:

```go
// Index by spec.nodeName for Pods
if err := mgr.GetFieldIndexer().IndexField(
    ctx,
    &corev1.Pod{},
    "spec.nodeName",
    func(obj client.Object) []string {
        pod := obj.(*corev1.Pod)
        if pod.Spec.NodeName == "" {
            return nil
        }
        return []string{pod.Spec.NodeName}
    },
); err != nil {
    return err
}
```

Two things about that call are worth making explicit, because the tidy example hides
them behind a convention.

**The index name is arbitrary.** That second argument, `"spec.nodeName"`, is only a
string key the index is registered under. `controller-runtime` does not parse it as
JSONPath and does not check it against the object's schema — you could write `"by-node"`
or `"xyzzy"` and it would behave identically. The only rule is that the *exact same
string* comes back in `MatchingFields` at query time. Naming the index after the field it
happens to read is a readability convention, nothing more.

**The indexed value is computed, not read.** The function returns whatever strings you
build; they need not be the verbatim contents of any single field. You can lowercase a

> value, join several fields into one composite key, bucket a timestamp (the **time-bucket**

Reviewer 4: ```suggestion
value, join several fields into one composite key, bucket a timestamp (the _time-bucket_
```

trick below does exactly this), or emit a string that appears nowhere in the object
literally. Whatever the function returns becomes a key in the inverted dictionary, and a
`MatchingFields` lookup for that exact key is what finds the objects again. The only
constraint is that the value has to be *derivable from the object you are indexing*.

> What is an **inverted index**? The term comes from search engines. Normally you have

Reviewer 4: nit: use italics when introducing new technical terms

)
```

This is not "fetch the full list, then filter". It is a lookup in the inverted index → a
ready set of keys → a fetch of the corresponding objects. A different code path entirely.

The comparison to SQL is more accurate than it might look at first:

| SQL | controller-runtime |
|---|---|
| `CREATE INDEX idx_node ON pods(node_name)` | `IndexField(&Pod{}, "spec.nodeName", fn)` |
| `SELECT * FROM pods WHERE node_name = 'node-1'` | `List(&pods, MatchingFields{"spec.nodeName": "node-1"})` |
| `SELECT * FROM obj WHERE owner_uid = $1` | `List(&list, MatchingFields{"metadata.ownerReferences.uid": uid})` (requires an `IndexField` for that field) |

Note the last row: `MatchingFields` does **not** make magic out of thin air. For every field
you want to look up via `MatchingFields` you need a corresponding `IndexField` registered
during manager setup. Without one, `controller-runtime` rejects the query and returns
an error.

A few things worth keeping in mind:

- **Equality only.** No range queries, no `LIKE`, no sorts, no aggregates. If you need
  "everything older than five minutes", either do a regular `List` and filter in code, or

>   use a **time-bucket** trick: instead of indexing the precise `time.Time`, index a rounded

Reviewer 4: nit: either use italics or no emphasis, not bold

  so common, there must be an optimization for them. There is not — `ThreadSafeStore` keeps
  no separate label dictionary.

  When you call `List(..., MatchingLabels{...})`, the controller honestly walks **every**
  cached object of the given type and checks each one against the selector. That is `O(n)`,
  exactly what `IndexField` is supposed to save you from.

  The API server itself supports filtering the event stream by a specific label selector. To
  make that effective in your controller, you have to optimize at the **cache population**
  stage, via `cache.ByObject{Label: ...}`, not at the **read** stage. This is covered in the
  next section on selective caches.

  And if you need a fast lookup by a specific label across already cached objects, register
  an `IndexField` for that label by hand. That works.
- **An index costs memory.** Every index is an extra dictionary keyed by every object. Do
  not index everything in sight speculatively.
- **You can only index data that is in the object itself.** You cannot index a Pod by "has a
  related PVC with such-and-such flag". Either store that bit in the Pod itself, or index
  the PVC, not the Pod.

> > **Note:** An index is built at registration time and is populated as part of the initial

Reviewer 4: This shouldn't be a block quote.

> `LIST`. By the time the first `Reconcile` runs, both `Get` and `List` with `MatchingFields`
> work correctly — the index is not built lazily.

## Selective cache: do not pull the whole cluster into your operator

By default, an informer pulls every object of its type from every namespace. For Pod,
Secret, ConfigMap, and Event in a large cluster, that is a multi-gigabyte surprise

> delivered on the first `LIST` at startup.

Reviewer 4: ```suggestion
delivered on the first **list** at startup.
```

It hurts especially with:

- **Secrets**, because Helm stores release state in them (`helm.sh/release.v1.*`), and those

>   secrets are easily a hundred kilobytes each.

Reviewer 2: Kubernetes style guide suggests avoiding words like “simply”, “just”, and “easily” where they do not add meaning, there are many words in this blog like that.
https://kubernetes.io/docs/contribute/style/style-guide/#avoid-words-that-assume-a-specific-level-of-understanding

Author: Good point — swept the whole article for "simply" / "just" / "easily" / "obvious(ly)": all removed or reworded. The one survivor is a temporal "just called" inside the quoted reconciler monologue, where it means "a moment ago" rather than assuming ease.

                Namespaces: map[string]cache.Config{
                    "my-operator": {},
                },
                Label: labels.SelectorFromSet(labels.Set{
                    "app.kubernetes.io/managed-by": "my-operator",
                }),
            },
            // Cache all Pods, but trim noise on the way into the store
            &corev1.Pod{}: {
                Transform: func(obj any) (any, error) {
                    pod := obj.(*corev1.Pod)
                    pod.ManagedFields = nil
                    return pod, nil
                },
            },
        },
    },
})
```

A subtle point: this is a **manager-level** setting and it affects **every controller in the
process** that reads the corresponding type. If you narrow the cache for Secrets to a single
namespace and another controller in the same binary needs all secrets in the cluster, that
controller will not see them. Before you tighten the scope, audit who else is reading
the type.

A short tour of the options:

- **`Namespaces`** restricts the visible scope. If your operator only manages its own
  namespace, there is no reason to keep other people's objects in memory.

> - **`Label` / `Field`** become parameters of the `WATCH` itself. The API server only sends

Reviewer 4: nit: write **watch** in bold lowercase

  matching objects, saving network and memory.
- **`Transform`** is invoked before the object lands in the store. It is the perfect place to
  drop `managedFields`, oversized `annotations`, or the binary `data` of ConfigMaps that
  you do not need.
- **`DefaultLabelSelector` / `DefaultNamespaces`** apply the same restriction globally, when
  every type needs the same scope.

> > **Caveat:** A selector limits what is **cached**, not what **exists**. If an object does

Reviewer 4: This shouldn't be a block quote.

> not match your selector, then as far as your operator is concerned, it does not exist in
> either `Get` or `List`. This bites people: somebody mislabels a single Secret and then
> spends half a day figuring out why their controller "cannot see it".

## Metadata-only: when `spec` and `data` are not needed

A separate pattern: you need to know that an object exists, but you do not need its `spec` or
`data`. Typical examples: a controller that waits for a Secret with a particular name to
appear but never reads it; one that counts PersistentVolume objects by the
`topology.kubernetes.io/zone` label; one that reacts to ConfigMap objects in a namespace
by name and does not care about contents.

> > **Caveat:** `PartialObjectMetadata` by definition gives you nothing from `spec` or

Reviewer 4: This shouldn't be a block quote.

the cache. When you actually need it:

- **Validating webhooks**, where the freshness of the object is critical. The cache in
  another process may be lagging at that very moment, and you would block a legitimate
  `Update`.
- A one-off read of a resource for which you do not maintain an informer. Spinning up a watch
  for a single operation is expensive.
- Reads **before `mgr.Start()`**, for instance during initialization. The regular
  `mgr.GetClient()` returns nothing useful at that point.
- **Paginated traversal of large result sets** through `client.Limit` / `client.Continue`.
  The cache-backed client ignores those parameters and always returns the full result set
  from the in-memory store; to actually page through the API server, you need `APIReader` (or
  a direct client of your own).

The price is a real network request. One thing to avoid: do not build "look in the cache, and
if missing, fall back to the API" logic. That is exactly the split-brain pattern the cache
is meant to protect you from.

### Disabling the cache for a type entirely

If you do not need a local cache for a given type at all — say, the type is "fat", read

> rarely, and the `LIST+WATCH` overhead is not worth paying — you can tell the manager not to

Reviewer 4: try to reword; it's not a huge problem, but HTTP has no verb `LIST+WATCH`.

cache it. This is configured through `client.Options.Cache.DisableFor`:

```go
mgr, err := ctrl.NewManager(cfg, ctrl.Options{
    Client: client.Options{
        Cache: &client.CacheOptions{
            DisableFor: []client.Object{
                &corev1.Secret{},
            },
        },
    },
})
```

With this configuration, `mgr.GetClient().Get(...)` and `List(...)` for Secret go straight
to the API server, bypassing the cache. No informer is started for that type, which means no

> `LIST` at startup and no permanent memory pressure from a store. This is a more radical

Reviewer 4: ```suggestion
`LIST` at startup and no permanent memory pressure from a store. This is a more radical
```

alternative to `APIReader`: where `APIReader` is reached for ad hoc, individual requests,
`DisableFor` turns the cache off for the type wholesale.

Real-world projects use this. Several established CNCF operators disable caching on Secrets,
both to save memory and to avoid hammering the API server with a large `LIST` at startup.

> > **Aside:** If you want to avoid a watch on the API server entirely, you can feed the

Reviewer 4: This shouldn't be a block quote.

> controller events from a source of your own design, bypassing `LIST+WATCH`. In
> `controller-runtime` this is done with `WatchesRawSource` / `source.Channel`: you can wire
> the controller to events from any place — an internal queue, a kubelet, a custom watch.
> Niche, but a perfectly valid pattern when the API server should not be touched.

> ## Best practices

Reviewer 4: ```suggestion
## Good practices
```

  store is empty, and you will get either `NotFound` or an empty `List` and then spend half
  a day investigating why an object "disappeared".
- **For deferred actions, use `RequeueAfter`,** not `time.Sleep` and not your own goroutines.

## Wrapping up

In one breath:

- The cache in `controller-runtime` is not an optimization, it is the operating model. Under
  the hood it is `Reflector` + `DeltaFIFO` + `Indexer` — exactly the same primitives that
  power Kubernetes itself.
- `r.Get` and `r.List` go to memory; `Create`, `Update`, `Patch`, and `Delete` go straight
  to the API server. Feedback flows in through the watch.
- `IndexField` plus `MatchingFields` turn the cache into a near-complete query engine with
  inverted indexes.
- `Namespaces`, selectors, `PartialObjectMetadata`, and `Transform` are the levers that
  control how much memory and traffic you actually consume.
- `APIReader` is the emergency exit for cases where you genuinely need the freshest version
  of an object.

And the single sentence to remember: `r.Get` inside a reconciler does not call the API
server. Ever. Not even the first time. Once that becomes a reflex, half the questions on

> operator code reviews answer themselves.

Reviewer 4: ```suggestion
controller code reviews answer themselves.
```

---
layout: blog
title: "How the controller-runtime Cache Actually Works, and Why Your Controller Does Not Crash the API Server"
draft: true
slug: controller-runtime-cache-explained
author: >
  Andrei Kvapil (Ænix),
  Timofei Larkin (Ænix)
---

Kubernetes has long been the default platform for distributed workloads, and writing your own

> operator for it is now a matter of a few hours. The standard path — `kubebuilder` on top of

Reviewer 4: ```suggestion
controller for it is now a matter of a few hours. The common path — Golang, using `kubebuilder` on top of
```

if any of them are already familiar.

- **GVK (GroupVersionKind)** — the triple that uniquely identifies a resource type in
  Kubernetes: group, version, and kind, for example `apps/v1/Deployment`. Almost every API in
  `controller-runtime` works in terms of GVK rather than the name you would type in `kubectl`
  (such as `deployments`).

- **resourceVersion** — a string the API server attaches to every object, marking the version
  it was last modified at (under the hood, a monotonically increasing counter from the storage
  layer). Clients pass it back unmodified; since Kubernetes 1.35 resource versions are
  [guaranteed to be orderable](/docs/reference/using-api/api-concepts/#resource-versions) as
  monotonically increasing integers within a single resource type (a conformance requirement).
  It serves two purposes: it powers
  **optimistic concurrency control** (on `Update`, the API server checks that your
  `resourceVersion` matches the one in `etcd`, otherwise it returns `409 Conflict`); and it is
  used to **resume a watch** (with `WATCH?resourceVersion=X`, the API server replays every

>   event that happened after version X).

Reviewer 4: Would this be OK?
```suggestion
- **resourceVersion** — a [monotonic counter](/docs/reference/using-api/api-concepts/#resource-versions)
  that the API server tracks automatically, and that gets increased every time an object gets updated.
  Although it's a decimal number, the field is represented as a string.
  Resource versions serve two main purposes. The first one, you can use them for
  *optimistic concurrency control* (for example: on **update**, the API server checks
  that the `resourceVersion` you provide matches the one in `etcd`, otherwise it
  returns `409 Conflict`). The second place you are going to see resource versions
  in your controller is to resume a **watch**. Read
  [watch bookmarks](/docs/reference/using-api/api-concepts/#watch-bookmarks) to learn
  more.
```

> - **Manager** — the `ctrl.Manager` object in `controller-runtime`. This is what your operator

Reviewer 4: ```suggestion
- **Manager** — the `ctrl.Manager` object in `controller-runtime`. This is what your controller
```

  enqueued at all (for example, "react only to changes in `spec`, ignore `status`").

With those in hand, you can dive in.

## Anatomy: what lives under the cache package

If you peek into `sigs.k8s.io/controller-runtime/pkg/cache`, you will see that it is a thin
wrapper over `k8s.io/client-go/tools/cache`. The same primitives that power the rest of
Kubernetes live underneath:

- **Reflector** — keeps a **watch** open against the API server and writes incoming changes
  into a queue as _deltas_. A delta is a record of the form "object X received an `Added` /
  `Updated` / `Deleted` event, and here is its new version". Effectively a single line in a
  change log.

- **DeltaFIFO** — the queue that holds those deltas. Per `namespace/name` key it accumulates
  the list of things that happened to that object, in order.

- **Indexer (Store)** — the in-memory object store, plus the indexes built over it.

- **SharedIndexInformer** — the conductor that ties everything together and dispatches events
  to subscribers — your controllers and any other observers.

At a glance the pipeline looks like this:

> {{< figure src="pipeline.svg" alt="Pipeline diagram: API server to Reflector to DeltaFIFO to Indexer to Event handlers" >}}

Reviewer 4: nit: that sounds like a caption, not alt text

Yes — at heart it is a map in memory. No B-trees, no LSMs. That is precisely why a cache-hit
`r.Get` costs microseconds: it is a map lookup followed by a copy of a Go struct.

### SharedIndexInformer and subscriptions

A SharedIndexInformer fuses Reflector, DeltaFIFO, and Indexer together and exposes two
interfaces to the rest of the world:

- Read objects directly from the indexer.
- Register a `ResourceEventHandler` and receive notifications for every event coming out of
  DeltaFIFO — `OnAdd`, `OnUpdate`, `OnDelete`. The store is updated in lockstep with the
  handler call, so by the time your handler runs, the indexer already reflects the new state.

"Outside" here means your controllers. When a controller registers `Watches(...)`, under the
hood it asks the informer: "add a handler that, on every change, enqueues the key into my
workqueue". The controller's workers then pop keys one at a time and call your
`Reconcile(ctx, ctrl.Request{NamespacedName: ...})`.

The keyword in the name is **Shared**. The manager creates **one** informer per GVK, and
every controller, webhook, and event source within that manager subscribes to it:

> {{< figure src="shared-informer.svg" alt="Shared informer diagram: a single LIST/WATCH per GVK feeding multiple subscribers" >}}

Reviewer 4: ```suggestion
{{< figure src="shared-informer.svg" caption="Shared informer diagram: a single list / watch per GVK, feeding multiple subscribers" >}}
```
You can also add alt text. Imagine you are talking to someone who cannot see very well and you are describing the diagram.

mutate the object and call `r.Update(ctx, &obj)`. The request goes to the API server right
now, and the API server checks:

- Does the `resourceVersion` in your PUT match the `resourceVersion` in `etcd`? Yes — write
  it.
- No, `etcd` already has a newer one? Reply with `409 Conflict` — somebody beat you to it.

This is _optimistic concurrency control_. No real locks are taken; everybody writes in
parallel; but only one of the racing `Update` calls wins — the one that arrives with the
current version. Everyone else gets a `409` and is expected to re-read and try again.

Why does this matter for the cache? If you naively send a PUT with "your" `resourceVersion`
from the cache and somebody has updated the object since you read it, you will get `409`.
That is **not a bug**. It is exactly the protection the system is supposed to give you.
Writing without the `resourceVersion` check (via `Patch` without an optimistic lock, or via
[Server-Side Apply](/docs/reference/using-api/server-side-apply/)) is also possible, but that
is a separate conversation.

The "write → visibility" cycle now looks like this:

> {{< figure src="update-visibility.svg" alt="Write visibility diagram: client.Update to API server to watch event to cache" >}}

Reviewer 4: ```suggestion
{{< figure src="update-visibility.svg" caption="Write visibility diagram: client.Update to API server to watch event to cache" >}}
```
You can also add alt text. Imagine you are talking to someone who cannot see very well and you are describing the diagram.

A real `relist` happens only in two cases: when the **watch** died with `410 Gone`, and when
you explicitly recreate the informer.

### Mistake 4: do not confuse `RequeueAfter` with a timer

A small note that often saves time. Sometimes you want to wait inside a reconciler — "we
just called the provider's API; if it is not ready yet, retry in a minute". The temptation is
to spin up `time.Sleep` or your own goroutine.

Resist it. `controller-runtime` already provides a built-in mechanism:

```go
return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
```

The controller puts your `req` back into the workqueue with a delayed trigger 30 seconds out.
If a real event for the same object arrives within that window, the reconcile fires
immediately, without waiting for the timer (the key is deduplicated in the queue). This is
both cheaper and more correct than a hand-rolled timer: you do not hold a worker, and you do
not risk missing a real event.

There is also `ctrl.Result{Requeue: true}` — enqueue immediately, subject to the rate
limiter.

## cache + index = almost SQL

Now you get to what is, arguably, the most useful capability of the cache — and the one most

> operators leave unused.

Reviewer 4: ```suggestion
controllers leave unused.
```

  cached object of the given type and checks each one against the selector. That is `O(n)`,
  exactly what `IndexField` is supposed to save you from.

  The API server itself supports filtering the event stream by a specific label selector. To
  make that effective in your controller, you have to optimize at the **cache population**
  stage, via `cache.ByObject{Label: ...}`, not at the **read** stage. This is covered in the
  next section on selective caches.

  And if you need a fast lookup by a specific label across already cached objects, register
  an `IndexField` for that label by hand. That works.
- **An index costs memory.** Every index is an extra dictionary keyed by every object. Do
  not index everything in sight speculatively.
- **You can only index data that is in the object itself.** You cannot index a Pod by "has a
  related PVC with such-and-such flag". Either store that bit in the Pod itself, or index
  the PVC, not the Pod.

**Note:** An index is built at registration time and is populated as part of the initial
**list**. By the time the first `Reconcile` runs, both `Get` and `List` with `MatchingFields`
work correctly — the index is not built lazily.

> ## Selective cache: do not pull the whole cluster into your operator

Reviewer 4: ```suggestion
## Selective cache: do not pull the whole cluster into your controller
```

You shouldn't assume that the reader intends to write an operator.

                },
            },
        },
    },
})
```

A subtle point: this is a **manager-level** setting and it affects **every controller in the
process** that reads the corresponding type. If you narrow the cache for Secrets to a single
namespace and another controller in the same binary needs all secrets in the cluster, that
controller will not see them. Before you tighten the scope, audit who else is reading
the type.

A short tour of the options:

- **`Namespaces`** restricts the visible scope. If your operator only manages its own
  namespace, there is no reason to keep other people's objects in memory.
- **`Label` / `Field`** become parameters of the **watch** itself. The API server only sends
  matching objects, saving network and memory.
- **`Transform`** is invoked before the object lands in the store. It is the perfect place to
  drop `managedFields`, oversized `annotations`, or the binary `data` of ConfigMaps that
  you do not need.
- **`DefaultLabelSelector` / `DefaultNamespaces`** apply the same restriction globally, when
  every type needs the same scope.

**Caveat:** A selector limits what is **cached**, not what **exists**. If an object does

> not match your selector, then as far as your operator is concerned, it does not exist in

Reviewer 4: ```suggestion
not match your selector, then as far as your controller is concerned, it does not exist in
```
